src/components/media-views.js

```diff
@@ -837,3 +846,150 @@ AFRAME.registerComponent("media-image", {
     this.el.emit("image-loaded", { src: this.data.src, projection: projection });
   }
 });
+
+AFRAME.registerComponent("media-pdf", {
+  schema: {
+    src: { type: "string" },
+    projection: { type: "string", default: "flat" },
+    contentType: { type: "string" },
+    index: { default: 0 },
+    batch: { default: false }
+  },
+
+  init() {},
+
+  remove() {
+    if (this.data.batch && this.mesh) {
+      this.el.sceneEl.systems["hubs-systems"].batchManagerSystem.removeObject(this.mesh);
+    }
+
+    if (this.currentPageTextureIsRetained) {
+      const cacheKey = this.currentTextureCacheKey();
+      textureCache.release(cacheKey);
+      this.currentPageTextureIsRetained = false;
+    }
+  },
+
+  currentTextureCacheKey() {
+    // We ensure a unique texture for each PDF entity, because they are drawn over during pagination.
+    return `${this.el.object3D.uuid}_${this.canvas.width}_${this.canvas.height}`;
+  },
+
+  async update(oldData) {
+    let texture;
+    let ratio = 1;
+
+    try {
+      const { src, index } = this.data;
+      if (!src) return;
+
+      this.el.emit("pdf-loading");
+
+      if (!oldData.src || src !== oldData.src) {
+        const loadingSrc = this.data.src;
+        this.pdf = await pdfjs.getDocument(src);
+        if (loadingSrc !== this.data.src) return;
+        this.el.setAttribute("media-pager", { maxIndex: this.pdf.numPages - 1 });
+      }
+
+      const loadingIndex = index;
+      const page = await this.pdf.getPage(index + 1);
+      if (loadingIndex !== this.data.index) return;
+
+      const viewport = page.getViewport({ scale: 3 });
+      const pw = viewport.width;
+      const ph = viewport.height;
+
+      if (!this.canvas || this.canvas.width !== pw || this.canvas.height !== ph) {
+        if (this.canvas) {
+          const oldCacheKey = `${this.el.object3D.uuid}_${this.canvas.width}_${this.canvas.height}`;
+
+          if (this.mesh.map !== errorTexture && textureCache.has(oldCacheKey)) {
+            textureCache.release(oldCacheKey);
+            this.currentPageTextureIsRetained = false;
+          }
+        }
+
+        this.canvas = document.createElement("canvas");
+        this.canvasContext = this.canvas.getContext("2d");
+        this.canvas.width = pw;
+        this.canvas.height = ph;
+      }
+
+      const renderingSrc = this.data.src;
+      const renderingIndex = index;
+
+      if (this.renderTask) {
```

Comment:
RenderTask has a `cancel` method. https://mozilla.github.io/pdf.js/api/draft/RenderTask.html, if you in fact still want to wait its probably better to just await on `this.renderTask.promise` instead of generating this new polling promise to wait on.

Reply:
ah yeah, I tried doing this at first. the reason this is difficult to do in another way is because the user can potentially slam on the pagination and this results in `update()` being called many times in succession, while a render is in-flight. to properly handle cancelling, you need to cancel and then wait for the original promise to resolve. so the orchestration here is quite complicated: you need to submit cancel, and then the originating call stack and your new call stack are both waiting on the same promise, and then if another request comes in, you don't want to re-cancel but still need to wait on the promise.

i was ready to wave the white flag until i realized I can just serialize the render requests, and the last one submitted will always be the one we want anyway, which is what it does now.

Reply:
i'll take another stab at it real quick

Reply:
OK, so I managed to avoid the interval and redundant renders by changing the control flow-- but cancelling seems problematic. If I cancel and handle it, it seems to corrupt the rendering somehow where the page ends up rendering into a small region on the canvas. Odds are it's something I'm doing wrong, but the net effect now is that the worst case is you need to wait for a single render to complete before seeing your page.